source/game/system/RaceManager.cc

```diff
@@ -265,7 +286,64 @@ void RaceManager::Player::decrementLap() {
 /// @addr{0x805349B8}
 void RaceManager::Player::incrementLap() {
     m_maxKcp = 0;
-    ++m_currentLap;
+    if (++m_currentLap <= m_maxLap) {
+        return;
+    }
+
+    const auto *kart = Kart::KartObjectManager::Instance()->object(0);
+    u16 addMs = CourseMap::Instance()->getCheckPointEntryOffsetMs(m_checkpointId, kart->pos(),
+            kart->prevPos());
+
+    const Timer &currentTimer = RaceManager::Instance()->timerManager().currentTimer();
+
+    Timer timer;
+    timer.mil = static_cast<u16>(static_cast<f32>(addMs) + static_cast<f32>(currentTimer.mil));
+
+    if (timer.mil > 999) {
+        // The additional milliseconds caused an overflow
+        u8 seconds = currentTimer.sec;
+        timer.mil -= 1000;
+        timer.sec = seconds + 1;
+
+        u16 addMin = 0;
+        if (timer.sec > 59) {
+            timer.sec -= 60;
+            addMin = 1;
+        }
+
+        timer.min = addMin + currentTimer.min;
+        if (timer.min > 999) {
+            timer.min = 999;
+            timer.sec = 59;
+            timer.mil = 999;
+        }
+    } else {
+        timer.sec = currentTimer.sec;
+        timer.min = currentTimer.min;
+    }
+
+    timer.valid = true;
+    Timer &lapTimer = m_lapTimers[m_maxLap - 1];
+    lapTimer.min = timer.min;
+    lapTimer.sec = timer.sec;
+    lapTimer.mil = timer.mil;
+    lapTimer.valid = true;
+
+    if (m_maxLap >= 3) {
+        endRace(timer);
+    } else {
+        m_maxLap = m_currentLap;
+    }
+}
+
+/// @addr{0x805347F4}
+void RaceManager::Player::endRace(const Timer &finishTime) {
+    m_raceTimer.min = finishTime.min;
+    m_raceTimer.sec = finishTime.sec;
+    m_raceTimer.mil = finishTime.mil;
+    m_raceTimer.valid = finishTime.valid;
```

Comment:
Default copy assignment operator should work fine
```suggestion
    m_raceTimer = finishTime;
```